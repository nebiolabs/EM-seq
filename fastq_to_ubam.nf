#!/usr/bin/env nextflow
nextflow.enable.dsl=2

input_glob = params.input_glob ?: ['*.{1,2}.fastq.gz']
read_format = params.read_format ?: 'paired-end'
params.outdir = './ubam'

process FastqToBamPaired {
    conda "bioconda::picard=3.3.0 bioconda::samtools=1.21"
    publishDir "${params.outdir}", mode: 'copy'
    memory { params.max_memory ?: 300.GB }
    
    input:
        tuple val(library), path(read1), path(read2)
    
    output:
        path('*.bam')

    shell:
    '''
    set +o pipefail

    flowcell=$(zcat !{read1} | head -n 1 | cut -d ":" -f 3)
    barcode=$(zcat !{read1} | head -n 1 | cut -d ":" -f 10)

    set -o pipefail

    picard FastqToSam TMP_DIR=/state/partition1/sge_tmp F1=!{read1} F2=!{read2} OUTPUT=/dev/stdout SM=!{library} LB=!{library} CN="New England Biolabs" PU=Illumina QUIET=true COMPRESSION_LEVEL=0 \
    | samtools view -h - \
    | while read -r line; do if [[ "$line" =~ ^@RG ]]; then printf "%s\\t%s\\n" "$line" "BC:${barcode}"; else printf "%s\\n" "$line"; fi; done \
    | samtools view -hb -o !{library}.bam -
    '''
}

process FastqToBamSingle {
    conda "bioconda::picard=3.3.0 bioconda::samtools=1.21"
    publishDir "${params.outdir}", mode: 'copy'
    memory { params.max_memory ?: 300.GB }
    
    input:
        path(read1)
    
    output:
        path('*.bam')

    shell:
    '''
    set +o pipefail

    library=$(basename !{read1} .fastq.gz)
    flowcell=$(zcat !{read1} | head -n 1 | cut -d ":" -f 3)
    barcode=$(zcat !{read1} | head -n 1 | cut -d ":" -f 10)

    picard FastqToSam F1=!{read1} OUTPUT=/dev/stdout SM=${library} LB=${library} CN="New England Biolabs" PU=Illumina QUIET=true COMPRESSION_LEVEL=0 \
    | samtools view -h - \
    | while read -r line; do if [[ "$line" =~ ^@RG ]]; then printf "%s\\t%s\\n" "$line" "BC:${barcode}"; else printf "%s\\n" "$line"; fi; done \
    | samtools view -hb -o ${library}.bam -
    '''
}

workflow {

    if (read_format == 'paired-end') {
        fastq_files = Channel.fromFilePairs(input_glob, flat: true)
        FastqToBamPaired(fastq_files)
    }
    else if (read_format == 'single-end') {
        fastq_files = Channel.fromPath(input_glob)
        FastqToBamSingle(fastq_files)
    }
    else {
        error "Unknown read format -- accepted is 'paired-end' or 'single-end'"
    }
}
