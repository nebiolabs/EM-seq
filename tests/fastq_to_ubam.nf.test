nextflow_pipeline {

    name "Test fastq to uBam"
    script "fastq_to_ubam.nf"
    tag "pipeline"

    test("fastq to uBam workflow - paired-end") {
        when {
            params.input_glob = "$projectDir/tests/fixtures/fastq/emseq-test*{.ds.1,.ds.2}.fastq.gz"
        }

        then {
            def ubam = bam("${launchDir}/ubam/emseq-test1.bam").getStatistics()

            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.trace,
                    ["ubam", ubam],
                ).match()
            }
        )
        }
    }

    test("fastq to uBam workflow - single-end") {
        when {
            params.input_glob = "$projectDir/tests/fixtures/fastq/emseq-test*.ds.1.fastq.gz"
            params.read_format = 'single-end'
        }

        then {
            def ubam = bam("${launchDir}/ubam/emseq-test1.ds.1.bam").getStatistics()

            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.trace,
                    ["ubam", ubam],
                ).match()
            }
        )
        }
    }
}
