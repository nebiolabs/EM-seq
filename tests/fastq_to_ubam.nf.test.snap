{
    "fastq to uBam workflow - single-end": {
        "content": [
            {
                "tasksFailed": 0,
                "tasksCount": 2,
                "tasksSucceeded": 2
            },
            [
                "ubam",
                {
                    "maxReadLength": 150,
                    "minReadLength": 150,
                    "meanReadLength": 150,
                    "maxQuality": 0,
                    "minQuality": 0,
                    "meanQuality": 0,
                    "readCount": 2500
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.2"
        },
        "timestamp": "2025-08-20T09:23:01.410361919"
    },
    "fastq to uBam workflow - paired-end": {
        "content": [
            {
                "tasksFailed": 0,
                "tasksCount": 2,
                "tasksSucceeded": 2
            },
            [
                "ubam",
                {
                    "maxReadLength": 150,
                    "minReadLength": 150,
                    "meanReadLength": 150,
                    "maxQuality": 0,
                    "minQuality": 0,
                    "meanQuality": 0,
                    "readCount": 5000
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.2"
        },
        "timestamp": "2025-08-20T09:22:42.948496009"
    }
}